core/src/main/scala/com/raphtory/internals/management/PythonEncoder.scala

```diff
@@ -181,10 +197,116 @@ object PythonEncoder {
 }
 
 object PythonInterop {
+  val logger: WrappedLogger = new WrappedLogger(Logger(LoggerFactory.getLogger(this.getClass)))
 
-  def assignId(s: String): Long =
+  def assign_id(s: String): Long =
     GraphBuilder.assignID(s)
 
+  def camel_to_snake(s: String): String =
+    PythonEncoder.camelToSnakeCase(s)
+
+  def testArgs(args: Any): Unit =
+    println(s"testArgs($args)")
+
+  def decode[T](obj: Any): T =
+    (obj match {
+      case obj: java.util.ArrayList[_] => obj.asScala
+      case obj                         => obj
+    }).asInstanceOf[T]
+
+  def get_wrapper_str(obj: Any): String =
+    obj match {
+      case _: Iterable[_]       => "Iterable"
+      case _: Iterator[_]       => "Iterator"
+      case _: GraphPerspective  => "TemporalGraph"
+      case _: Table             => "Table"
+      case _: Accumulator[_, _] => "Accumulator"
+      case _: Vertex            => "Vertex"
+      case _: GraphState        => "GraphState"
+      case _                    => "None"
+    }
+
+  def find_class(name: String): Any = {
+    val runtimeMirror = universe.runtimeMirror(getClass.getClassLoader)
+    val module        = runtimeMirror.staticModule(name)
+    val obj           = runtimeMirror.reflectModule(module)
+    obj.instance
+  }
+
+  def make_varargs(obj: Iterable[Any]): Array[Object] =
+    obj.toArray.map(_.asInstanceOf[Object])
+
+  def methods(obj: Any): util.Map[String, Array[Method]] = {
+    logger.trace(s"Scala 'methods' called with $obj")
+    val prefixedMethodDict         = mutable.Map.empty[String, mutable.ArrayBuffer[java.lang.reflect.Method]]
+    val prefixedMethodDefaultsDict = mutable.Map.empty[String, mutable.Map[Int, java.lang.reflect.Method]]
+    obj.getClass.getMethods.foreach { m =>
+      val parts: Array[String] = """\$default\$""".r.split(m.getName)
+      val prefix               = camel_to_snake(parts(0))
+
+      if (parts.length == 1)
+        prefixedMethodDict.getOrElseUpdate(prefix, mutable.ArrayBuffer.empty[java.lang.reflect.Method]).append(m)
+      else {
+        val defaultIndex = parts(1).toInt - 1
+        prefixedMethodDefaultsDict
+          .getOrElseUpdate(prefix, mutable.Map.empty[Int, java.lang.reflect.Method])
+          .addOne(defaultIndex, m)
+      }
+    }
+    val res                        = prefixedMethodDict
+      .map {
+        case (name, methods) =>
+          val defaults = prefixedMethodDefaultsDict.get(name) match {
+            case Some(v) => v.toMap
+            case None    => Map.empty[Int, java.lang.reflect.Method]
+          }
+
+          name -> methods.map { m =>
+            val hasVarArgs  = m.isVarArgs
+            val paramsNames = m.getParameters.map(p => camel_to_snake(p.getName))
+
+            val n = m.getParameterCount
+            if (
```

Comment:
Needs an explanation of why we are checking if method has non default arguments. Also a point on inplicits.

Reply:
I've added some comments to explain what this is doing. It may be worth revisiting this method to see if we can use scala reflection instead to simplify things. Implicit arguments do not appear to be special from the java side.